server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java

```diff
@@ -120,6 +120,9 @@ public void readBytes(byte[] b, int offset, int len) {
 
     @Override
     public int read(byte[] b, int off, int len) throws IOException {
+        if (available() <= 0) {
+            return -1;
+        }
         int toRead = Math.min(len, available());
```

Comment:
Also let's avoid calling `available()` twice, this is on a hot enough path for that to matter I think